src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,433 @@
+            /// }
+            /// After:
+            /// void Caller(bool x)
+            /// {
+            ///     DoSomething(Foo(), Bar());
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// In this case, parameters 'a' and 'b' should just be replaced by the argument expression.
+            /// Note: this might cause semantics changes. It is by design.
+            /// </summary>
+            public ImmutableDictionary<IParameterSymbol, TExpressionSyntax> ParametersToReplace { get; }
+
+            public MethodParametersInfo(
+                ImmutableArray<(IParameterSymbol parameterSymbol, string name)> parametersWithVariableDeclarationArgument,
+                ImmutableArray<(IParameterSymbol parameterSymbol, TExpressionSyntax expression)> parametersToGenerateFreshVariablesFor,
+                ImmutableDictionary<IParameterSymbol, TExpressionSyntax> parametersToReplace)
+            {
+                ParametersWithVariableDeclarationArgument = parametersWithVariableDeclarationArgument;
+                ParametersToGenerateFreshVariablesFor = parametersToGenerateFreshVariablesFor;
+                ParametersToReplace = parametersToReplace;
+            }
+        }
+
+        private async Task<MethodParametersInfo> GetMethodParametersInfoAsync(
+            Document document,
+            TMethodDeclarationSyntax calleeMethodNode,
+            IInvocationOperation invocationOperation,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var allArgumentOperations = invocationOperation.Arguments
+                .WhereAsArray(argument => argument.Children.IsSingle());
+
+            // 1. Find all the parameter maps to an identifier from caller. After inlining, this identifier would be used to replace the parameter in callee body.
+            // For params array, it should be included here if it is accept an array identifier as argument.
+            // Example:
+            // Before:
+            // void Caller(int i, int j, bool[] k)
+            // {
+            //     Callee(i, j, k);
+            // }
+            // void Callee(int a, int b, params bool[] c)
+            // {
+            //     DoSomething(a, b, c);
+            // }
+            // After:
+            // void Caller(int i, int j, bool[] k)
+            // {
+            //     DoSomething(i, j, k);
+            // }
+            // void Callee(int a, int b, params bool[] c)
+            // {
+            //     DoSomething(a, b, c);
+            // }
+            var operationsWithIdentifierArgument = allArgumentOperations
+                .WhereAsArray(argument =>
+                    _syntaxFacts.IsIdentifierName(argument.Children.Single().Syntax) && argument.ArgumentKind == ArgumentKind.Explicit);
```

Comment:
note: this could change semantics.  in the case taht the identifier maps to a Property, and now you're executing the property in a different order than before.

Just doc that we know this can happen, but it is very unlikely, and would indicate very strange code, and this is a refactoring anyways, so we accept it.

Reply:
LOL, looks like a similar case we discussed :)